### ISSUE TITLE
How to register a ligand in the ForceField class which is covalently bound to an amino acid

### ISSUE THREAD

**Issue by ctx-mebert** [openmm/openmm#4084]
I would like to model a covalently modified lysine which has a ligand attached to it (example see image below). I have tried using the template generator of GAFF but due to the modification the lysine is not recognized anymore. I also tried registering the entire lysine+ligand via GAFF and renaming the modified lysine from LYS -> LYM but then the previous N-1 residue is not recognized anymore. Here code snippets of what I tried:

```python
from openff.toolkit.topology import Molecule
from openmmforcefields.generators import GAFFTemplateGenerator
molecule = Molecule.from_file('nnaa.sdf')
gaff = GAFFTemplateGenerator(molecules=molecule)
forcefield = ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
forcefield.registerTemplateGenerator(gaff.generator)
pdbfluoro = PDBFile('lysozyme-fluoro.pdb')
modellerfluoro = Modeller(pdbfluoro.topology, pdbfluoro.positions)
modellerfluoro.addMembrane(forcefield)
```

Error if ligand is part of the lysine residue (error is in the N-1 position not the lysine position itself):
```
 raise ValueError('No template found for residue %d (%s).  %s  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template' % (res.index+1, res.name, _findMatchErrors(self, res)))
ValueError: No template found for residue 47 (ASP).  The set of atoms matches ASP, but the bonds are different.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template
```

Error if ligand is in a residue itself covalently bound to the lysine residue:
```
    raise ValueError('No template found for residue %d (%s).  %s  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template' % (res.index+1, res.name, _findMatchErrors(self, res)))
ValueError: No template found for residue 48 (LYS).  The set of atoms matches LYN, but the bonds are different.  For more information, see https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template
```

Thank you for your help and tips on how to register a covalently modified amino acid / non-natural amino acid.

<img width="395" alt="image" src="https://github.com/openmm/openmm/assets/117865920/0d65e01c-9499-4ea8-bc6d-8321f948077c">

**Comment by peastman** [comment]
You probably just need to add CONECT records for the modified residue, including the peptide bonds connecting it to the ones surrounding it.  PDB files omit CONECT records for standard residues, but nonstandard ones need to have every bond listed.  See [the FAQ](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template) for more information about this.

**Comment by ctx-mebert** [comment]
So the CONECT record to N-1 and N+1 got rid of the issue with the previous residue not being detected. However, I can't get the modified residue right for the SDF file to be detected in the force field. I attached the lysozyme example with the LYX in the zip and the SDF I made from it. I am not sure why it is not the correct hit mask. The example protein is non scientific to place in a membrane but it is just for testing. How would I need to define an amino acid in the SDF to be detected?

[Archive.zip](https://github.com/openmm/openmm/files/11576052/Archive.zip)

**Comment by peastman** [comment]
OpenMM doesn't read SDF files.  What are you doing with it?

I recommend putting the entire molecule in a single PDB file.  It should include the modified amino acid plus the ligand attached to it, all listed as a single residue.  Every bond within that residue should be explicitly listed.

**Comment by j-wags** [comment]
Hi @ctx-mebert - The presence of a sdf file here makes me wonder if you're intending to use OpenFF's [modified residue example](https://gist.github.com/Yoshanuikabundi/66007cb9966b1455a259baaf7cd7e7c3)? There's also a [newer-but-worse-documented version](https://github.com/openforcefield/2023-workshop-vignettes/tree/main#2023-openff-workshop-vignettes) using some early functionality prototypes that we showed in our recent workshop. If you have questions about this, you can contact us on the Open Force Field [GitHub Discussions page](https://github.com/orgs/openforcefield/discussions).

**Comment by ctx-mebert** [comment]
@peastman I am trying to add parameters to an existing forcefield in OpenMM via the GAFFTemplateGenerator from openmmforcefields which I attach to the ForceField object from openmm. The molecule topology object for the GAFFTemplateGenerator is provided via an SDF file read by openff.toolkit.topology.

@j-wags indeed I am trying to implement this suggestion to dynamically add parameters for covalently bound ligands and also wanted to experiment in general with non-natural amino acids. Do you have any workflow for this?

Thanks for all your help here.

**Comment by j-wags** [comment]
Hi @ctx-mebert, the first [modified residue example I linked above](https://gist.github.com/Yoshanuikabundi/66007cb9966b1455a259baaf7cd7e7c3) is OpenFF's current best solution - it shows how to use OpenFF to assign parameters to a modified protein using our port of FF14SB for the canonical amino acids, and parameters from OpenFF's Sage force field for a non-canonical amino acid. This workflow isn't something we officially support as its own produce, but we used it as an example to show what users can build with our toolkits and would probably solve your problem. It is completely independent of the OpenMMForceFields package and the OpenMM ForceField class, instead using the OpenFF ForceField class to load SMIRNOFF-format force fields, which can combine force fields this way. 

One important questions to ask if you choose to pursue the OpenMMForceFields path is how you will assign parameters at the interface of the standard and modified parts of the protein. While per-atom parameters like vdW can come entirely from the protein or small molecule force field, what about torsions that include some atoms in the protein and others in the small molecule? Outside of clever atom naming magic to make sure that all the parameters can be looked up from the force fields (which I don't think OpenMMForceFields does, though I'm not an expert with it), it's not clear to me how the interface-spanning parameters would be handled by OpenMM/OpenMMForceFIelds.
